File: src/mvc/cpp/item_delegate/custom_delegate.cpp
#include "custom_delegate.h"
#include <QLineEdit>
#include <QPainter>
#include <QProgressBar>
#include <QDialog>
#include <QPushButton>
#include <QVBoxLayout>
#include <QApplication>

CustomDelegate::CustomDelegate(QObject *parent) : QStyledItemDelegate(parent)
{

}


CustomDelegate::~CustomDelegate()
{

}


void CustomDelegate::paint(QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index) const
{
    auto row = index.row();

    switch (row) {
    case 1:{ 
        // create QProgressBar
        auto progressBar = new QProgressBar();
        progressBar->setGeometry(option.rect);
        progressBar->setValue(index.data().toInt());
        progressBar->setTextVisible(false);
        progressBar->setStyleSheet("QProgressBar::chunk { background-color: #006400; }");
        QPixmap pixmap(option.rect.size());
        progressBar->render(&pixmap);
        painter->drawPixmap(option.rect, pixmap);
        break;
    }
    case 3:{
        auto oldPen = painter->pen();
        QPen pen(QColor("#FFD700"), 1, Qt::DashLine, Qt::RoundCap, Qt::RoundJoin);
                    painter->setPen(pen);
                                        painter->setBrush(QColor("#FFD700"));
                                        painter->setRenderHint(QPainter::Antialiasing);

    painter->drawRect(option.rect.center().x() - 30, option.rect.center().y() -10, 20, 20);
    painter->drawRect(option.rect.center().x() + 10, option.rect.center().y() -10, 20, 20);
    painter->setPen(oldPen);

        break;
    }
    default: {
     // draw background blue
        painter->fillRect(option.rect, QColor(176,224,230));
        // ceter text
        auto text = index.data().toString();
        auto rect = option.rect;
        auto textRect = painter->boundingRect(rect, Qt::AlignCenter, text);
        textRect.moveCenter(rect.center());
        painter->drawText(textRect, Qt::AlignCenter, text);
        return;
    }
    }

}


QSize CustomDelegate::sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const
{
    // make it twice as long as the default
    auto originalSize = QStyledItemDelegate::sizeHint(option, index);
    return QSize(originalSize.width() , originalSize.height() * 2);
}


QWidget *CustomDelegate::createEditor(QWidget *parent, const QStyleOptionViewItem &option, const QModelIndex &index) const
{
    auto row = index.row();
    switch (row) {
    case 1: {
        auto dialog = new QDialog(parent);
        dialog->setWindowFlags(Qt::Popup);
        auto slider = new QSlider(Qt::Horizontal);
        slider->setRange(0, 100);
        slider->setValue(index.data().toInt());
        slider->setTickPosition(QSlider::TicksBothSides);
        slider->setTickInterval(10);
        dialog ->setLayout(new QVBoxLayout);
        dialog->layout()->addWidget(slider);
        return dialog;
    }
    case 3: {
        return nullptr;
    }
    default: {
          auto editor = new QLineEdit(parent);
        editor->setText(index.data().toString());
        return editor;
    }
    }
}

void CustomDelegate::setModelData(QWidget *editor, QAbstractItemModel *model, const QModelIndex &index) const
{

    auto row = index.row();
    switch (row) {
    case 1: {
        auto dialog = qobject_cast<QDialog*>(editor);
        auto slider = qobject_cast<QSlider*>(dialog->findChild<QSlider*>());
        model->setData(index, slider->value());
        return;
    }
    default: {
        auto lineEdit = qobject_cast<QLineEdit*>(editor);
        model->setData(index, lineEdit->text());
        return;
            }
    }

}

File: src/mvc/python/item_delegate/custom_delegate.py
from PyQt6.QtCore import QSize, Qt
from PyQt6.QtGui import QColor, QPainter, QPen, QPixmap
from PyQt6.QtWidgets import (
    QDialog,
    QLineEdit,
    QProgressBar,
    QSlider,
    QStyledItemDelegate,
    QVBoxLayout,
)


class CustomDelegate(QStyledItemDelegate):
    def __init__(self, parent=None):
        super().__init__(parent)

    def paint(self, painter, option, index):
        row = index.row()

        if row == 1:
            # create QProgressBar
            progress_bar = QProgressBar()
            progress_bar.setGeometry(option.rect)
            progress_bar.setValue(int(index.data()))
            progress_bar.setTextVisible(False)
            progress_bar.setStyleSheet(
                "QProgressBar::chunk { background-color: #006400; }"
            )
            pixmap = QPixmap(option.rect.size())
            progress_bar.render(pixmap)
            painter.drawPixmap(option.rect, pixmap)
        elif row == 3:
            old_pen = painter.pen()
            pen = QPen(
                QColor("#FFD700"),
                1,
                Qt.PenStyle.DashLine,
                Qt.PenCapStyle.RoundCap,
                Qt.PenJoinStyle.RoundJoin,
            )
            painter.setPen(pen)
            painter.setBrush(QColor("#FFD700"))
            painter.setRenderHint(QPainter.RenderHint.Antialiasing)

            painter.drawRect(
                option.rect.center().x() - 30, option.rect.center().y() - 10, 20, 20
            )
            painter.drawRect(
                option.rect.center().x() + 10, option.rect.center().y() - 10, 20, 20
            )
            painter.setPen(old_pen)
        else:
            # draw background blue
            painter.fillRect(option.rect, QColor(176, 224, 230))
            # center text
            text = index.data()
            rect = option.rect
            text_rect = painter.boundingRect(rect, Qt.AlignmentFlag.AlignCenter, text)
            text_rect.moveCenter(rect.center())
            painter.drawText(text_rect, Qt.AlignmentFlag.AlignCenter, text)

    def sizeHint(self, option, index):
        # make it twice as long as the default
        original_size = QStyledItemDelegate.sizeHint(self, option, index)
        return QSize(original_size.width(), original_size.height() * 2)

    def createEditor(self, parent, option, index):
        row = index.row()
        if row == 1:
            dialog = QDialog(parent)
            dialog.setWindowFlags(Qt.WindowType.Popup)
            slider = QSlider(Qt.Orientation.Horizontal)
            slider.setRange(0, 100)
            slider.setValue(int(index.data()))
            slider.setTickPosition(QSlider.TickPosition.TicksBothSides)
            slider.setTickInterval(10)
            dialog.setLayout(QVBoxLayout())
            dialog.layout().addWidget(slider)
            return dialog
        elif row == 3:
            return None
        else:
            editor = QLineEdit(parent)
            editor.setText(index.data())
            return editor

    def setModelData(self, editor, model, index):
        row = index.row()
        if row == 1:
            dialog = editor
            slider = dialog.findChild(QSlider)
            model.setData(index, slider.value())
        else:
            line_edit = editor
            model.setData(index, line_edit.text())
